### PR TITLE
When using `Start-Transcript` and file exists, empty file rather than deleting

### LINKED ISSUE #4065: Start-Transcript might silently delete the output file
If Start-Transcript is ran without -Append and a user has write privileges on the file without write privileges on the parent folder, the file will be silently deleted without any kind of error message.

Steps to reproduce
------------------

As administrator create a folder and file with the privileges specified above. I'm assuming that a local user, "myuser" exist:
```powershell
mkdir "c:\test\"
icacls.exe "c:\test" "/grant" "builtin\administrators:(CI)(OI)(F)" "/inheritance:r"
icacls.exe "c:\test" "/grant" "myuser:(CI)(OI)(RX)"
"" | Out-File -Encoding "ascii" -LiteralPath "c:\test\my-transcript.txt"
icacls.exe "c:\test\my-transcript.txt" "/grant:r" "myuser:(F)"
```

Run Start-Transcript as "myuser":
```powershell
Start-Transcript -LiteralPath "c:\test\my-transcript.txt"
Stop-Transcript
```

Expected behavior
-----------------

The transcript should be written to c:\test\my-transcript.txt

Actual behavior
---------------

The output states success, but the file is actually deleted.

> PS C:\> Start-Transcript -LiteralPath "c:\test\my-transcript.txt"
> Transcript started, output file is c:\test\my-transcript.txt
> 
> PS C:\> Stop-Transcript
> Transcript stopped, output file is C:\test\my-transcript.txt
> 
> PS C:\> Test-Path "c:\test\my-transcript.txt"
> False

Environment data
----------------
Not the newest PowerShell ...
> PS C:\> $PSVersionTable
> 
> Name                           Value                                                                                                                                                                                                                                       
> ----                           -----                                                                                                                                                                                                                                       
> PSVersion                      5.1.14393.1358                                                                                                                                                                                                                              
> PSEdition                      Desktop                                                                                                                                                                                                                                     
> PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                                                     
> BuildVersion                   10.0.14393.1358                                                                                                                                                                                                                             
> CLRVersion                     4.0.30319.42000                                                                                                                                                                                                                             
> WSManStackVersion              3.0                                                                                                                                                                                                                                         
> PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                                         
> SerializationVersion           1.1.0.1     

... but you can clearly see the error in line 230 of StartTranscriptCmdlet.cs :
https://github.com/PowerShell/PowerShell/blob/02b5f357a20e6dee9f8e60e3adb9025be3c94490/src/Microsoft.PowerShell.ConsoleHost/host/msh/StartTranscriptCmdlet.cs#L230

```c#
// If they didn't specify -Append, delete the file
if (!_shouldAppend)
{
	System.IO.File.Delete(effectiveFilePath);
}
```

This part of the code should rather write some sort of zero length content to the file.

### PR DESCRIPTION
**Pull request by paalbra**
## PR Summary

Currently Start-Transcript will delete the file it refers to if it's not set to append. This might lead to issues if the user running the command only has access to the file and not the parent folder. This PR makes sure the file is emptied rather than deleted. There are more details in the issue. Fix #4065. 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

## P.S.

This is my first PR in this project. I'm hopefully doing things right...

### CODE REVIEW

**Comment by iSazonov** [comment]
My understanding is that currently we write an error and remove the file. Your suggestion is  don'nt remove the file but wipe. Right?

/cc @mklement0 What do you think about the special case?

**Comment by paalbra** [comment]
> My understanding is that currently we write an error and remove the file. Your suggestion is don'nt remove the file but wipe. Right?

Well, kind of. The file is deleted and when Start-Transcript tries to recreate the file you get an access denied error (if the access is set in a way mentioned in #4065 ). But it is a bit OS/filesystem dependent.

This can be avoided by not deleting the file, but rather wiping it.

The fix should not alter the behavior of successful usage of Start-Transcript, but it will fix this special access case.

**Comment by mklement0** [comment]
I too agree that we shouldn't special-case this:

`Start-Tanscript` makes no promises about preexisting files; it blindly tries to replace them, and always creates the new files with the default permissions.

In other words: loss of any preexisting file with the same name is always to be expected and no relationship between the old and the new file can and should be assumed.

That creation of a _new_ file then fails, is a separate issue, and requires that the permissions be set accordingly on the enclosing directory to allow file creation by the current user - which strikes me as reasonable.

@paalbra: Your scenario sounds exotic and the steps to reproduce it (as shown in #4065) are quite involved. Is this a real-life scenario you've run into, and why would you expect to run into this repeatedly?

**Comment by iSazonov** [comment]
@mklement0 Thanks!

@SteveL-MSFT Please confirm that current behavior is "by-design" and we can not accept the PR.

**Comment by paalbra** [comment]
Yes, I've run into this issue in my real-life environment. I don't view this issue as a special or exotic case myself.

You could even compare this to Out-File which also has an Append parameter. Out-File has no issues with the same access permissions set because Out-File doesn't delete the file. I would expect both commands to work in this setup and this is what I try to fix with this PR.

**Comment by SteveL-MSFT** [comment]
@paalbra I appreciate you bringing up this inconsistent experience and taking the time to submit a PR, however, we need to balance changes with impact to backwards compatibility.  In this specific case, it seems the current behavior is documented so it's not a strong enough justification to change the current behavior.  (I'm always open to reconsidering if there's sufficient community feedback)

**Comment by paalbra** [comment]
With the current behavior you also have other bad side effects. E.g. you can't point Start-Transcript to a symlink. If you do Start-Transcript will just delete the symlink and create a file.

There is simply no reason for Start-Transcript to delete the file. I should just overwrite the file. This is also what it says that it does in the documentation. Overwriting a file is not the same as recreating a file. So I think the current documentation and behavior don't match.

**Comment by paalbra** [comment]
I really can't agree that the behavior is the same as the documentation.

Again you could compare Start-Transcript and Out-File. They both have similar documentation when it comes to Append and NoClobber, but they do not behave the same way.

**Comment by mklement0** [comment]
@paalbra: 

I see your point now: `Out-File` (without `-Append`)  / `>` and `Set-Content` indeed open a preexisting file and _truncate_ it before writing the new content.
`>` in POSIX-like shells works the same.

In light of this, I think your PR has merit.

@SteveL-MSFT: 

Unless I'm missing something, I think this PR would be an easy consistency improvement that shouldn't impact existing scripts.

**Comment by SteveL-MSFT** [comment]
@paalbra I misunderstood the intent, but thanks to @mklement0 I get it.  Based on the explanation, I would agree that this PR makes sense particularly in the symlink case.

**Comment by SteveL-MSFT** [comment]
@paalbra it would be great if you can add a test for this specific behavior.  You can probably just check the file creation date hasn't changed.

**Comment by paalbra** [comment]
@SteveL-MSFT I'm not very familiar with the project yet, since this is my first PR, but I'll certainly look into adding more tests.

**Comment by iSazonov** [comment]
@paalbra  Test is here https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.Powershell.Host/Start-Transcript.Tests.ps1

**Comment by SteveL-MSFT** [comment]
@paalbra Just give it a try and we'll make suggestions :)

**Comment by paalbra** [comment]
A bit more of searching the web led me to the [FileSystemWatcher ](https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher?view=netframework-4.7.2) class.

I guess I could use one of those and make a test that makes sure no file deletion events are emitted when running Start-Transcript. This seems a bit excessive to me though. Any thoughts?

**Comment by SteveL-MSFT** [comment]
@paalbra Thanks for making the effort.  I agree that using FileSystemWatcher is overkill for this.  I suspect you can't reply on the `CreationTime` because the filesystem or disk driver hasn't flushed it before the write happens within that window.  Another option (on Windows, perhaps skip on non-Windows) is to remove `Delete` permission from that file and check that no error is returned (see `Set-Acl`)

**Comment by paalbra** [comment]
I decided to try the FileSystemWatcher path since I got it working on Windows and Ubuntu (the test is failing in master and passing in my branch).

But for some reason it's failing on MacOS. I'm really not sure why and I currently have no Mac available for debugging. I'll try to get my hands on a Mac some day soon though...

**Comment by SteveL-MSFT** [comment]
@paalbra there may be an issue in CoreFx where the file on macOS is deleted or that the event is being incorrectly raised

**Comment by vexx32** [comment]
@paalbra looks like the file gets deleted on MacOS when you're not expecting it to, at least from the build logs:
```
2018-10-30T18:20:10.2962680Z TEST FAILURES
2018-10-30T18:20:10.2962740Z Description: Should not delete the file if it already exist
2018-10-30T18:20:10.2963680Z Name:        Start-Transcript, Stop-Transcript tests.Should not delete the file if it already exist
2018-10-30T18:20:10.2963770Z message:
2018-10-30T18:20:10.2964240Z Expected $null, but got 'FileDeleted'.
2018-10-30T18:20:10.2964700Z stack-trace:
2018-10-30T18:20:10.2965270Z at <ScriptBlock>, /Users/vsts/agent/2.141.1/work/1/s/test/powershell/Modules/Microsoft.Powershell.Host/Start-Transcript.Tests.ps1: line 127
2018-10-30T18:20:10.2965760Z 127:         Receive-Job $job | Should -Be $null
2018-10-30T18:20:10.2965830Z 1 tests in test/powershell failed
2018-10-30T18:20:10.2965890Z At /Users/vsts/agent/2.141.1/work/1/s/build.psm1:1347 char:13
2018-10-30T18:20:10.2966380Z +             throw "$($x.'test-results'.failures) tests in $TestArea f ...
2018-10-30T18:20:10.2966450Z +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2018-10-30T18:20:10.2966990Z + CategoryInfo          : OperationStopped: (1 tests in test/powershell failed:String) [], RuntimeException
2018-10-30T18:20:10.2967050Z + FullyQualifiedErrorId : 1 tests in test/powershell failed
```

**Comment by mklement0** [comment]
I have no explanation, but here's a data point: I just ran the test locally against this PR's code on my macOS 10.14 machine, and it succeeded in 100 successive invocations.

**Comment by iSazonov** [comment]
How we open the file? If "Read shared" we could open the file twice and check its content by second handle.

**Comment by mklement0** [comment]
Interesting, @paalbra, but there's some consistency: I hadn't used the `test:` drive in my tests.

**Comment by SteveL-MSFT** [comment]
Perhaps to help isolate if this is a `$TestDrive` issue or something due to PSDrives, can you try creating a PSDrive and see if you still get delete events?

> New-PSDrive -PSProvider FileSystem -Name MyTestDrive -Root $env:TMPDIR

**Comment by paalbra** [comment]
Ok, so I've done some more testing. On MacOS the TestDrive doesn't seem to be "clean" between tests. Somehow the state/file deletion events "bleed through" from other tests. I've tried to resolve this by creating a new, unique path within my test and this seems to work. `$transcriptFilePath = Join-Path $TestDrive ([System.IO.Path]::GetRandomFileName())`

Any thoughts or feedback on this method?

**Comment by paalbra** [comment]
I guess this is something you might want for all tests. So a better solution might be to create a new unique path in the `BeforeAll` block. I'm not sure I should touch that block in this PR though. It might be better to create this as an separate issue?

**Comment by mklement0** [comment]
I think creating a unique path for this PR is definitely the way to go.

My sense is that we needn't worry about the other tests, as they don't use file-system watchers.

It is curious that only macOS is affected, though. 

There are several open issues for the `FileSystemWatcher` type in the CoreFx repo, though only 
 one macOS-specific one and one cross-platform one: https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+filesystemwatcher++in%3Atitle

Another thing to try - though more out of curiosity, given that the use `[System.IO.Path]::GetRandomFileName()` seems like the best solution - is to call CLI `sync` beforehand, which commits pending disk writes system-wide - note that this can be expensive.

**Comment by SteveL-MSFT** [comment]
@paalbra thanks for taking the time to investigate this.  My suggestion is for this PR to have the macOS specific code path (with appropriate comments).  Agree with @mklement0 that the other tests are probably ok although not ideal.  It may be worthwhile to see if `sync` helps which would indicate to me that this is an OS problem with caching.

**Comment by paalbra** [comment]
Adding `sync` doesn't seem to help. I've also tried adding sleep like this:

```powershell
It "Should not delete the file if it already exist" {
    # Create an existing file
    #$transcriptFilePath = Join-Path $TestDrive ([System.IO.Path]::GetRandomFileName())
    Out-File $transcriptFilePath
    Start-Sleep -Seconds 20

    $FileSystemWatcher = [System.IO.FileSystemWatcher]::new((Split-Path -Parent $transcriptFilePath), (Split-Path -Leaf $transcriptFilePath))

    $Job = Register-ObjectEvent -InputObject $FileSystemWatcher -EventName "Deleted" -SourceIdentifier "FileDeleted" -Action {
        return "FileDeleted"
    }

    Start-Transcript -Path $transcriptFilePath
    Stop-Transcript

    Unregister-Event -SourceIdentifier "FileDeleted"

    # Nothing should have been returned by the FileSystemWatcher
    Receive-Job $job | Should -Be $null
}
```

The above actually makes the test pass on MacOS.

My guess is that the `FileSystemWatcher` receives some kind of old events. Maybe events from the previous `AfterEach` block where the file is removed.

I think the current solution, with unique file path, is the best for all OSes.

**Review by iSazonov — APPROVED** [review]

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Only need to address @iSazonov's comment on try..finally

File: test/powershell/Modules/Microsoft.Powershell.Host/Start-Transcript.Tests.ps1 [hunk 1]
```diff
@@ -108,6 +108,25 @@ Describe "Start-Transcript, Stop-Transcript tests" -tags "CI" {
...
+        $FileSystemWatcher = [System.IO.FileSystemWatcher]::new((Split-Path -Parent $transcriptFilePath), (Split-Path -Leaf $transcriptFilePath))
+
+        $Job = Register-ObjectEvent -InputObject $FileSystemWatcher -EventName "Deleted" -SourceIdentifier "FileDeleted" -Action {
+            return "FileDeleted"
+        }
+
+        Start-Transcript -Path $transcriptFilePath
+        Stop-Transcript
+
+        Unregister-Event -SourceIdentifier "FileDeleted"
```
[the remark this answers is not in the document]
Comment (paalbra) [reply test/powershell/Modules/Microsoft.Powershell.Host/Start-Transcript.Tests.ps1]: I pushed a commit doing this now

**Review by SteveL-MSFT — APPROVED** [review]
LGTM . Thanks!

**Comment by iSazonov** [comment]
@paalbra Thanks for your contribution!

**Pull request merged**